### PR TITLE
Interactivity Updates for Tables and Maps

### PR DESCRIPTION
**Pull request by chrispyles**
[x] Wrote test for feature
[x] Added changes to CHANGELOG.md
[x] Bumped version number (delete if unneeded)

**Changes proposed:**
This PR includes new interactivity updates that allow for plotly-based plots in addition to their matplotlib defaults and new options for `Marker#map_table` and `Circle#map_table`. It also includes a Data 8 student-friendly reference notebook in the documentation.

### CODE REVIEW

**Comment by chrispyles** [comment]
Hey @davidwagner @adnanhemani this PR is failing due to the coverage decrease. Because all of the methods added aren't tested by pytest (since they're all plotting/mapping) what do you suggest be done so this check passes?

**Comment by davidwagner** [comment]
Adnan wrote some code to include everything covered by notebooks in the coverage numbers (PR #432), so that might be one clean solution, but I haven't figured out how to get that PR working on my end yet.

**Comment by davidwagner** [comment]
It looks like this introduces a bunch of new functionality: e.g., clusters, index maps, colorbar scales, radius units.  I'm guessing those are orthogonal to whether the maps are interactive vs not.  Can you talk a bit about that?  What's the motivation for adding that functionality?  It looks like could use some careful attention and I'm not sure I know this codebase well enough to render a thoughtful opinion on them.  If we introduce these arguments, will we have to continue to support them in the non-interactive plot methods, and should we expect that to impose a maintenance burden?

I also see a bunch of additional documentation, which looks useful independent of any interactivity changes.  Cool.

**Comment by davidwagner** [comment]
@adnanhemani Would you like the opportunity to review this PR?

File: datascience/tables.py
Comment (adnanhemani) [inline datascience/tables.py:2338]: No need for this. The _visualize method already has default params for this, unless I'm missing something...

File: datascience/tables.py
Comment (adnanhemani) [inline datascience/tables.py:2599]: also not necessary

File: datascience/tables.py
Comment (adnanhemani) [inline datascience/tables.py:2899]: Not necessary

File: datascience/tables.py
Comment (adnanhemani) [inline datascience/tables.py:4207]: Not necessary

File: docs/tables.rst
Comment (adnanhemani) [inline docs/tables.rst:140]: If our preferred way of users using the interactive plots is by using the interactive method, then I'm not sure that we should be documenting the methods straight. Rather, merge this documentation into the non-interactive method one's - and that shouldn't add much extra stuff, given that the method signatures should be identical.

**Comment by adnanhemani** [comment]
Making some comments here since diff is too large to show:
* The URL for world-countries.json needs to be changed back to the link. It's been made a local filename for some reason and it shouldn't be.
* When looking at plotly-charts.ipynb, under the `Working with Returned Figures` header, all the Figures are getting printed out. We need to suppress those.
* All the CSVs that were added in for the reference-notebook need to be turned into reading from URLs (I'm guessing you downloaded these from somewhere in CSV format). These files added an extra ~12MB to the entire repo and we try to keep the repo as light as possible. I'd say the same for the geocoding CSV, but that's still understandable as we want all our functionality to work offline, as much as possible.

Overall, to respond with my two cents to @davidwagner's thoughts...I think there's a bunch of new functionality and complication that we are introducing via this PR - but after reading through it all, I'm leaning on the side of saying it's worth it to have in our package. There will no doubt be extra maintenance overhead (heck, there's so much code for the interactive plots that I just skipped because I don't have the time nor mental capacity to digest all of it at the moment) but I think it can be manageable if @khsu2000 and @chrispyles can commit to taking up maintenance for all interactive plotting methods; I think I understand most of the new maps functionality to take responsibility for it if all the changes I asked for are made.

File: datascience/tables.py
Comment (chrispyles) [reply datascience/tables.py]: I think all of these are necessary... `Table#_visualize` specifies the defaults of the width and height as default values for kwargs but since it is called with these kwargs set to the values in the function (i.e. `self._visualize(..., width=width, height=height, ...)`) these get overwritten with `None`. Will put these into a class attribute so that we don't hardcode the same values multiple times but I think these statements are necessary.

File: datascience/tables.py
Comment (chrispyles) [reply datascience/tables.py]: fixed. Created `Table#_import_plotly` which gets called in all the right places.

File: docs/tables.rst
Comment (chrispyles) [reply docs/tables.rst]: changed.

File: datascience/tables.py
Comment (adnanhemani) [reply datascience/tables.py:2362]: ahh you're right, I was thinking of something else. good catch!

**Comment by chrispyles** [comment]
@adnanhemani @davidwagner OK I've pushed resolutions to some of these comments. Resolved the specific edits above and addressed all 3 of @adnanhemani's bullet-points above. @khsu2000 is working on the updates to maps, which we anticipate having completed by Friday.

File: datascience/maps.py
Comment (khsu2000) [reply datascience/maps.py]: Done, rewrote this using datascience tables only.

File: datascience/maps.py
Comment (khsu2000) [reply datascience/maps.py:903]: Done, rewrote this using datascience tables only.

File: datascience/maps.py
Comment (khsu2000) [reply datascience/maps.py]: Used self._attrs for these args instead, so that there's no new if statements in Map initializer.

File: datascience/maps.py
Comment (khsu2000) [reply datascience/maps.py:123]: This is for a new cluster_by feature that can take groups of markers and assign them each to a separate marker_cluster. In order to distinguish the groups, we needed to make new marker_cluster icons that include a mouse-over feature.

**Comment by chrispyles** [comment]
Added tests to get the coverage back to where it was @adnanhemani. @khsu2000 and I are meeting tomorrow to discuss the last edit but all else is gtg.

File: datascience/maps.py
Comment (chrispyles) [reply datascience/maps.py]: OK we've collapsed some of this information into a new set of kwargs called `map_kwargs` which gets passed into the `Map` constructor. @khsu2000 has some objections to passing around the original `kwargs` dictionary that he can elaborate on if needed.

**Comment by chrispyles** [comment]
OK @adnanhemani @davidwagner all changes have been made, tests added, etc. Should be ready to merge from our end; please let @khsu2000 and I know if anything else is needed.

Also re: maintenance, @khsu2000 and I are OK with taking responsibility for maintenance of the new plotly stuff and whatever else as needed.

**Comment by adnanhemani** [comment]
Hi @khsu2000 and @chrispyles - last thing on my end and then it looks great to me. @davidwagner would you also like to review to this make sure I haven't left anything out?

File: datascience/maps.py
Comment (chrispyles) [reply datascience/maps.py]: Just pushed a fix. `pkg_resources.resource_filename` basically allows us to find the path to a data file that is stored in the same location where the package is installed, so we can access the data file we need to construct the table.

**Comment by chrispyles** [comment]
Removed the `colors` argument from `Table#iscatter`. Let's see if travis passes.

**Review by adnanhemani — APPROVED** [review]

**Comment by adnanhemani** [comment]
LGTM - I'll let @davidwagner have final say before I merge this since it is quite large.

**Comment by davidwagner** [comment]
@adnanhemani   This looks good to me!  I have a small pending comment (about listing plotly twice in linux_environment.yml), but that should be easy to fix.  I think we're good to go to merge it.

Thank you for all your work on this, @chrispyles and @khsu2000!

**Comment by davidwagner** [comment]
I've merged this -- thank you for all the work, @chrispyles and @khsu2000!

I will rely on you in the future for any software maintenance or bug reports that touch these pieces.

**Pull request closed without merging**
